File: dotnet/src/Envdoctor/Runtime/Capture.cs
using System.Diagnostics;
using System.Runtime.InteropServices;
using System.Text.Json;
using System.Text.RegularExpressions;
using Envdoctor.Models;

namespace Envdoctor.Runtime;

/// Live-machine runtime snapshot capture. Captures the current machine's OS,
/// installed tool versions, `$PATH`, opt-in global package inventory, and the
/// *names* of non-secret environment variables. Secret-looking names are
/// dropped, never masked, and values are never captured.
public static class Capture
{
    /// Tools probed by default. Order here is the display order before sorting.
    private static readonly (string Tool, string[] Args)[] ToolProbes =
    {
        ("node", new[] { "-v" }),
        ("python3", new[] { "--version" }),
        ("python", new[] { "--version" }),
        ("go", new[] { "version" }),
        ("rustc", new[] { "-V" }),
        ("java", new[] { "-version" }),
        ("ruby", new[] { "-v" }),
        ("php", new[] { "-v" }),
        ("perl", new[] { "-v" }),
        ("cc", new[] { "--version" }),
        ("git", new[] { "--version" }),
    };

    private static readonly Regex VersionRe = new(@"(\d+\.\d+(?:\.\d+)?)", RegexOptions.Compiled);

    /// Collapse a leading `$HOME` to `~` so snapshots don't leak usernames and
    /// stay comparable across machines.
    private static string CollapseHome(string p)
    {
        var home = Environment.GetEnvironmentVariable("HOME") ?? "";
        if (home.Length > 0 && (p == home || p.StartsWith(home + "/", StringComparison.Ordinal)))
            return "~" + p[home.Length..];
        return p;
    }

    /// Ordered, de-duplicated `$PATH` entries with `$HOME` collapsed.
    private static List<string> CollectPath()
    {
        var raw = Environment.GetEnvironmentVariable("PATH") ?? "";
        var seen = new HashSet<string>();
        var result = new List<string>();
        foreach (var part in raw.Split(':'))
        {
            if (part.Length == 0)
                continue;
            var entry = CollapseHome(part);
            if (seen.Add(entry))
                result.Add(entry);
        }
        return result;
    }

    /// Non-secret env var NAMES only, sorted. Secret-looking names are dropped.
    private static List<string> CollectEnvFlagNames()
    {
        var names = Environment.GetEnvironmentVariables().Keys
            .Cast<object>()
            .Select(k => k.ToString() ?? "")
            .Where(name => !EnvironmentVariable.IsSecretName(name))
            .ToList();
        names.Sort(StringComparer.Ordinal);
        return names;
    }

    private static string? Run(string tool, string[] args, bool redirectErr = true)
    {
        try
        {
            var startInfo = new ProcessStartInfo(tool)
            {
                RedirectStandardOutput = true,
                RedirectStandardError = redirectErr,
                UseShellExecute = false,
            };
            foreach (var arg in args)
                startInfo.ArgumentList.Add(arg);
            using var process = Process.Start(startInfo);
            if (process is null)
                return null;
            var stdout = process.StandardOutput.ReadToEnd();
            var stderr = redirectErr ? process.StandardError.ReadToEnd() : "";
            process.WaitForExit();
            return stdout + stderr;
        }
        catch
        {
            return null;
        }
    }

    /// Probe one CLI's version; returns null when the tool isn't installed.
    private static string? ProbeVersion(string tool, string[] args)
    {
        var output = Run(tool, args);
        if (output is null)
            return null;
        var match = VersionRe.Match(output);
        return match.Success ? match.Groups[1].Value : null;
    }

    /// Locate which PATH directory a command resolves from, `$HOME` collapsed.
    private static string ResolveFrom(string tool)
    {
        var stdout = Run("which", new[] { tool }, redirectErr: false);
        if (stdout is null)
            return "";
        var first = stdout.Split('\n').FirstOrDefault(l => l.Trim().Length > 0);
        if (first is null)
            return "";
        var dir = Path.GetDirectoryName(first.Trim()) ?? "";
        return CollapseHome(dir);
    }

    /// Probe every known tool; only installed ones appear, sorted by name.
    private static List<ToolInfo> CollectTools()
    {
        var tools = ToolProbes
            .Select(probe => (probe.Tool, Version: ProbeVersion(probe.Tool, probe.Args)))
            .Where(r => r.Version is not null)
            .Select(r => new ToolInfo
            {
                Tool = r.Tool,
                Version = r.Version!,
                ResolvedFrom = ResolveFrom(r.Tool),
            })
            .ToList();
        tools.Sort((a, b) => string.CompareOrdinal(a.Tool, b.Tool));
        return tools;
    }

    /// Global package inventory, opt-in because it is slow. Best-effort per
    /// ecosystem (currently npm).
    private static Dictionary<string, List<GlobalPackage>> CollectGlobals()
    {
        var globals = new Dictionary<string, List<GlobalPackage>>();
        var stdout = Run("npm", new[] { "ls", "-g", "--depth=0", "--json" }, redirectErr: false);
        if (stdout is not null)
        {
            var pkgs = ParseNpmGlobals(stdout);
            if (pkgs.Count > 0)
                globals["npm"] = pkgs;
        }
        return globals;
    }

    /// Parse `npm ls -g --json` into a name/version list; tolerant of partial JSON.
    private static List<GlobalPackage> ParseNpmGlobals(string stdout)
    {
        try
        {
            using var doc = JsonDocument.Parse(stdout);
            if (!doc.RootElement.TryGetProperty("dependencies", out var deps) ||
                deps.ValueKind != JsonValueKind.Object)
            {
                return new List<GlobalPackage>();
            }
            var pkgs = deps.EnumerateObject()
                .Select(p => new GlobalPackage
                {
                    Name = p.Name,
                    Version = p.Value.TryGetProperty("version", out var v) && v.ValueKind == JsonValueKind.String
                        ? v.GetString() ?? ""
                        : "",
                })
                .ToList();
            pkgs.Sort((a, b) => string.CompareOrdinal(a.Name, b.Name));
            return pkgs;
        }
        catch (JsonException)
        {
            return new List<GlobalPackage>();
        }
    }

    /// OS release string (`uname -r` on unix), best-effort.
    private static string OsRelease()
    {
        var stdout = Run("uname", new[] { "-r" }, redirectErr: false);
        var trimmed = stdout?.Trim();
        return string.IsNullOrEmpty(trimmed) ? "unknown" : trimmed;
    }

    /// OS/arch strings matching Node's `process.platform` / `process.arch`
    /// (the reference CLI's snapshot shape).
    private static string OsPlatform()
    {
        if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
            return "darwin";
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
            return "linux";
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            return "win32";
        if (RuntimeInformation.IsOSPlatform(OSPlatform.FreeBSD))
            return "freebsd";
        return "unknown";
    }

    private static string OsArch() => RuntimeInformation.OSArchitecture switch
    {
        Architecture.X64 => "x64",
        Architecture.X86 => "ia32",
        Architecture.Arm64 => "arm64",
        Architecture.Arm => "arm",
        _ => "unknown",
    };

    /// Capture this machine's live runtime. `globals` opts into the slow
    /// package inventory.
    public static RuntimeSnapshot CaptureSnapshot(bool globals) =>
        new()
        {
            Schema = SnapshotSchema.Value,
            CapturedAt = DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString(),
            Os = new OsInfo
            {
                Platform = OsPlatform(),
                Arch = OsArch(),
                Release = OsRelease(),
            },
            Tools = CollectTools(),
            Path = CollectPath(),
            Globals = globals ? CollectGlobals() : new Dictionary<string, List<GlobalPackage>>(),
            EnvFlagNames = CollectEnvFlagNames(),
        };
}

File: elixir/lib/envdoctor/runtime/capture.ex
defmodule Envdoctor.Runtime.Capture do
  @moduledoc """
  Capture this machine's live runtime into a snapshot. Never records any
  environment variable **value** — only non-secret names in `envFlagNames`,
  and secret-looking names are dropped entirely.
  """

  alias Envdoctor.Scanner

  @snapshot_schema 1

  # First dotted version-looking token in a tool's output.
  @version_re ~r/(\d+\.\d+(?:\.\d+)?)/

  # Tools probed by default. Order here is irrelevant; results are sorted.
  @tool_probes [
    {"node", ["-v"]},
    {"python3", ["--version"]},
    {"python", ["--version"]},
    {"go", ["version"]},
    {"rustc", ["-V"]},
    {"java", ["-version"]},
    {"ruby", ["-v"]},
    {"php", ["-v"]},
    {"perl", ["-v"]},
    {"cc", ["--version"]},
    {"git", ["--version"]}
  ]

  @doc "Snapshot schema version; `snapshot-diff` refuses tokens from newer versions."
  def snapshot_schema, do: @snapshot_schema

  @doc "Capture the live runtime. Pass `globals: true` for the npm package inventory."
  def capture(opts \\ []) do
    tools = collect_tools()
    globals = if Keyword.get(opts, :globals, false), do: collect_globals(), else: %{}
    {platform, arch, release} = os_info()

    %{
      schema: @snapshot_schema,
      capturedAt: DateTime.utc_now() |> DateTime.truncate(:second) |> DateTime.to_iso8601(),
      os: %{platform: platform, arch: arch, release: release},
      tools: tools,
      path: collect_path(),
      globals: globals,
      envFlagNames: collect_env_flag_names()
    }
  end

  defp os_info do
    platform =
      case :os.type() do
        {:unix, :darwin} -> "darwin"
        {:unix, :linux} -> "linux"
        {:unix, :freebsd} -> "freebsd"
        {:win32, _} -> "win32"
        {:unix, other} -> Atom.to_string(other)
        {family, _} -> Atom.to_string(family)
      end

    arch =
      case :erlang.system_info(:system_architecture) |> to_string() |> String.split("-") do
        ["aarch64" | _] -> "arm64"
        ["x86_64" | _] -> "x64"
        [other | _] -> other
      end

    release =
      case :os.type() do
        {:unix, _} ->
          case System.cmd("uname", ["-r"], stderr_to_stdout: false) do
            {out, 0} -> String.trim(out)
            _ -> ""
          end

        _ ->
          ""
      end

    {platform, arch, release}
  rescue
    _ -> {platform_fallback(), "", ""}
  end

  defp platform_fallback do
    case :os.type() do
      {:unix, name} -> Atom.to_string(name)
      {family, _} -> Atom.to_string(family)
    end
  end

  @doc "Collapse a leading $HOME to \"~\" so snapshots don't leak usernames."
  def collapse_home(p) do
    case System.user_home() do
      nil ->
        p

      home ->
        if p == home or String.starts_with?(p, home <> "/") do
          "~" <> String.slice(p, byte_size(home)..-1//1)
        else
          p
        end
    end
  end

  @doc "Ordered, de-duplicated `$PATH` entries with $HOME collapsed. Order is significant."
  def collect_path(path_env \\ System.get_env("PATH") || "") do
    path_env
    |> String.split(":")
    |> Enum.reject(&(&1 == ""))
    |> Enum.map(&collapse_home/1)
    |> Enum.uniq()
  end

  @doc "Non-secret env var NAMES only. Secret-looking names are dropped, not masked."
  def collect_env_flag_names(env \\ System.get_env()) do
    env
    |> Map.keys()
    |> Enum.reject(&Scanner.secret_name?/1)
    |> Enum.sort()
  end

  @doc "Probe one CLI's version; nil when the tool isn't installed or misbehaves."
  def probe_version(tool, args) do
    with path when is_binary(path) <- System.find_executable(tool),
         {out, _} <- run_quietly(path, args),
         [_, version] <- Regex.run(@version_re, out) do
      version
    else
      _ -> nil
    end
  end

  defp run_quietly(executable, args) do
    task =
      Task.async(fn ->
        try do
          System.cmd(executable, args, stderr_to_stdout: true)
        rescue
          _ -> {"", 1}
        catch
          :exit, _ -> {"", 1}
        end
      end)

    case Task.yield(task, 4_000) || Task.shutdown(task, :brutal_kill) do
      {:ok, {out, _}} -> {out, 0}
      _ -> nil
    end
  end

  @doc "Locate which PATH directory a command resolves from, $HOME collapsed."
  def resolve_from(tool) do
    case System.find_executable(tool) do
      nil -> ""
      path -> path |> Path.dirname() |> collapse_home()
    end
  end

  @doc "Probe every known tool; only installed ones appear in the result."
  def collect_tools do
    @tool_probes
    |> Enum.flat_map(fn {tool, args} ->
      case probe_version(tool, args) do
        nil -> []
        version -> [%{tool: tool, version: version, resolvedFrom: resolve_from(tool)}]
      end
    end)
    |> Enum.sort_by(& &1.tool)
  end

  @doc "Global npm package inventory, opt-in (`--globals`) because it is slow."
  def collect_globals do
    case System.find_executable("npm") do
      nil ->
        %{}

      npm ->
        task =
          Task.async(fn ->
            try do
              System.cmd(npm, ["ls", "-g", "--depth=0", "--json"], stderr_to_stdout: false)
            rescue
              _ -> {"", 1}
            catch
              :exit, _ -> {"", 1}
            end
          end)

        case Task.yield(task, 15_000) || Task.shutdown(task, :brutal_kill) do
          {:ok, {out, 0}} ->
            case Envdoctor.Json.decode(out) do
              {:ok, %{"dependencies" => deps}} when is_map(deps) ->
                pkgs =
                  deps
                  |> Enum.map(fn {name, meta} ->
                    %{name: name, version: if(is_map(meta), do: Map.get(meta, "version", ""), else: "")}
                  end)
                  |> Enum.sort_by(& &1.name)

                if pkgs == [], do: %{}, else: %{npm: pkgs}

              _ ->
                %{}
            end

          _ ->
            %{}
        end
    end
  end
end
